src/shared/find-esp.c

```diff
@@ -283,21 +285,24 @@ static int verify_fsroot_dir(
                         if (r < 0)
                                 return log_error_errno(r, "Failed to extract parent path from '%s': %m", path);
 
-                        r = RET_NERRNO(stat(parent, &st2));
-                }
+                        r = get_block_device(parent, &devt_b);
```

Comment:
BTW, in what situation this fallback logic works??

Reply:
i have a testing setup here where /boot/ and /boot/efi are set up the way fedora does by default, i.e. /boot is a separate partition and /boot/efi is the ESP. If you are unprivileged you can enumerate the contents of /boot, but /boot/efi is made inaccessible via the dmask= mount option: you may stat() the /boot/efi inode itself but may not enumerate or access anything inside it. This means we /boot/ is accessible, but /boot/efi/.. is not...

hence this annyoing fallback.

I'll add a longer comment.

Reply:
Interesting, thank you for the explanation.